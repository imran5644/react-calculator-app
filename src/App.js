function App() {
  return (
    <div className="App">
     Calculator-app
    </div>
  );
}

export default App;
