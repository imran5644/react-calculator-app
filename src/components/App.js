import React, { Component } from 'react';
import ButtonPanel from './ButtonPanel';
import Display from './Display';
import calculate from '../logic/calculate';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      total: null,
      next: null,
      operation: null,
      lastClicked: null,
    };
  }

   handleClick = (buttonName) => {
    const { total, next, operation, lastClicked } = this.state;
    const operations = ['+', '-', 'รท', 'X', '='];
    
     if (buttonName === '.') {
      this.setState({
        next: next == null ? `0${buttonName}` : next + buttonName,
      });
    }
     else if (buttonName in ['1', '2', '3', '4', '5', '6', '7', '8', '9', '0']) {
       this.setState = ({
         next: next == null ? buttonName : next + buttonName,
       });
     } else {
       switch (buttonName) {
         case '+':
         case '-':
         case 'X':
         case 'รท':
         

           case '=':
            if (total === null || total === 'undefined') {
              this.setState({
                total: next,
                next: null,
                operation: buttonName,
              });
            } else if (!operations.includes(lastClicked)) {
           this.setState(calculate({ total, next, operation}, buttonName));
            }
           break;

         default:
          this.setState(calculate({ total, next, operation }, buttonName));
       }
     }
     this.setState({ lastClicked: buttonName });
   }

   render() {
     const { next, total } = this.state;
     return (
       <>
         <Display result={next === null ? total : next} />
         <ButtonPanel clickHandler={this.handleClick} />
       </>
     );
   }
}

export default App;
