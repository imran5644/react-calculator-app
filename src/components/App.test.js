import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

describe('Calculator', () => {
  it('Renders the Calculator page', () => {
    const calculator = renderer
    .create(<App />)
    .toJSON();
    expect(calculator).toMatchSnapshot();
  });
});

describe('Calculator component', () => {
  beforeEach(() => {
    render(<App />);
  })

  it('Adds two numbers', async () => {
    await userEvent.click(screen.getByText('3'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('+'));
    await userEvent.click(screen.getByText('2'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('='));
    const display = await screen.findByText(/500/);
    expect(display).toBeInTheDocument();
  });

  it('Substracts two numbers', async () => {
    await userEvent.click(screen.getByText('5'));
    await userEvent.click(screen.getByText('2'));
    await userEvent.click(screen.getByText('5'));
    await userEvent.click(screen.getByText('-'));
    await userEvent.click(screen.getByText('5'));
    await userEvent.click(screen.getByText('='));
    const display = await screen.findByText(/520/);
    expect(display).toBeInTheDocument();
  });

  it('Multiplies two numbers', async () => {
    await userEvent.click(screen.getByText('4'));
    await userEvent.click(screen.getByText('X'));
    await userEvent.click(screen.getByText('2'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('='));
    const display = await screen.findByText(/80/);
    expect(display).toBeInTheDocument();
  });

  it('Divides two numbers', async () => {
    await userEvent.click(screen.getByText('2'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('รท'));
    await userEvent.click(screen.getByText('2'));
    await userEvent.click(screen.getByText('='));
    const display = await screen.findByText(/10/);
    expect(display).toBeInTheDocument();
  });

  it('Turns number in percentage', async () => {
    await userEvent.click(screen.getByText('8'));
    await userEvent.click(screen.getByText('%'));
    const display = await screen.findByText(/0.08/);
    expect(display).toBeInTheDocument();
  });

  it('Turns positive number into negative number', async () => {
    await userEvent.click(screen.getByText('7'));
    await userEvent.click(screen.getByText('+/-'));
    const display = await screen.findByText(/-7/);
    expect(display).toBeInTheDocument();
  });

  it('Turns negative number into positive number', async () => {
    await userEvent.click(screen.getByText('7'));
    await userEvent.click(screen.getByText('6'));
    await userEvent.click(screen.getByText('+/-'));
    await userEvent.click(screen.getByText('+/-'));
    const display = await screen.findByText(/76/);
    expect(display).toBeInTheDocument();
  });

  it('Clears the display', async () => {
    await userEvent.click(screen.getByText('8'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('AC'));
    expect(screen.queryByText('80')).toBeNull();
  });  
})