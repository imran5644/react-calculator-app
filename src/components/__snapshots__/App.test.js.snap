// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator snapshot Renders the Calculator page 1`] = `
<main>
  <h1>
    Let's do some Math!
  </h1>
  <div
    className="calculator"
  >
    <div
      className="Display"
    >
      
    
    
    </div>
    <div>
      <button
        className="Button "
        name="AC"
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        className="Button "
        name="+/-"
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        className="Button "
        name="%"
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        className="operator "
        name="รท"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
    </div>
    <div>
      <button
        className="Button "
        name="7"
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        className="Button "
        name="8"
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        className="Button "
        name="9"
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="operator "
        name="X"
        onClick={[Function]}
        type="button"
      >
        X
      </button>
    </div>
    <div>
      <button
        className="Button "
        name="4"
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        className="Button "
        name="5"
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        className="Button "
        name="6"
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="operator "
        name="-"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
    </div>
    <div>
      <button
        className="Button "
        name="1"
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        className="Button "
        name="2"
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        className="Button "
        name="3"
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="operator "
        name="+"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
    </div>
    <div>
      <button
        className="Button "
        name="0"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        className="Button "
        name="."
        onClick={[Function]}
        type="button"
      >
        .
      </button>
      <button
        className="operator "
        name="="
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</main>
`;
